### ISSUE TITLE
Convert ElizaWy's Idle/Sit/Jump/Climb Assets Over

### ISSUE THREAD

**Issue by jrconway3** [sanderfrenken/Universal-LPC-Spritesheet-Character-Generator#141]
Eliza has many clothing, accessory, and hairstyle assets for Sit/Jump/Climb. Run won't work unless we fully replace the Run animation with Eliza's run, at least for the clothes. However, the hairstyles could be ported over easy enough.

I'll experiment with seeing how easy it is to create the classic "Male Run" animation based on ElizaWy's run. I suspect this may be quite difficult and this is a complex animation, but it'd be ideal if we could use this.

**Comment by jrconway3** [comment]
So I do have an update here now. I'm trying to work on mapping out these assets to the ULPC and I've covered all the clothing assets. Between "Male" and "Female" (which is equivalent to "Teen" and "Female" here), the only assets that are truly different are the torso assets, and I already have all "Teen" clothing assets mapped. Any "leg" or "feet" pieces I'll just make one universal set between female and teen.

There's four main torso assets:
1 - T-Shirt (this appears to match "Shortsleeve Shirt")
2 - Longsleeve Shirt (this appears to match ULPC's "Longsleeve Shirt")
3 - Cardigan (I don't see this, but its similar to Longsleeve Shirt)
4 - Polo (very similar to Longsleeve Shirt)

T-Shirt and Longsleeve Shirt are further divided into four types each:
- Normal
- V-Neck
- Scoop
- Buttoned

On ULPC, I do see Scoop Longsleeve Shirt, but that's the only variation I see.


Because the Longsleeve Shirt already exists on ULPC, it shouldn't be hard to grab the sleeves from these assets for the frames that are missing on Eliza's and merge them together. Eliza's assets don't have Cast (which is very easy for Leg assets because the legs don't move), Shoot, Thrust, and Down.


I'm really not worried about Hair at all, though. IIRC the hairstyles are already on ULPC anyway, and the only notable differences are that the Jump and Run hairstyles are animated... and that's about it. So these should be fairly easy to include.


I have noticed that the sixth "Jump" frame (the landing one, which is identical to the second Jump frame) has some slightly different animations for some assets. Personally speaking I didn't think the difference was noticeable enough to worry about it, but I could include the sixth Jump frame as well potentially just so the landing frame can have a slightly different animation but I don't think its a huge concern.

**Comment by jrconway3** [comment]
I put aside the clothing for now and switched over to hairstyles. I'll open a PR soon containing all hairstyles mapped over.

I think I may have made a small mistake, but it probably won't be a big deal? I mapped all these hairstyles only using female because I thought that it should be fine to only use one. However, I forgot that the reason why there's male and female separate is because some hairstyles try and put hair going "behind" the body but without any sort of layering. I have been using layering more often which is why I haven't had this problem myself.

I'll open this PR for now, but if it becomes a problem we may want to make this just for female/teen/pregnant and add separate male assets back.


For most of these hairstyles I combined the original hairstyle with Eliza's and recolored Eliza's to our palettes. However, there were a few hairstyles that did not exist on ULPC, so these will be new. They're similar to existing ones but still pretty unique; Curly (not the same as the current Curly Short/Long) as well as "Parted". Despite being called "Parted", its a totally different hairstyle, so I labeled it "Parted 2". (The original Parted is in CC-BY-SA, while Eliza only uses OGA-BY assets.)

Some of these hairstyles were redrawn and shrunk down to fit into the Revised smaller heads. In those cases I just dropped using Eliza's mapping all-together and simply remapped the existing assets on ULPC to all frames. Thankfully these particular ones were only short hairstyles mapped straight to the head (which is why they were resized) and as such there was no animations on these hairstyles, so I didn't have to worry about that particular issue at all.

**Comment by jrconway3** [comment]
All right, all hairstyles have been merged. Here's what's still missing:
- Facial Hair (I think this should be pretty easy, I can basically do this whenever)
- Eyebrows (pretty sure we don't have any eyebrows at all, so this would be a great addition)
- Head Accessories (Glasses, Eyepatch)
- Helmets (these might not work as they're probably mapped to the Revised smaller heads, but I'll check anyway)
- Clothing (I think I'll work on finishing clothing next tbh; I just need to make some standardized "base model" assets and then finish all the remaining arms)
- Bowtie is already done, I just haven't included it yet. I kinda wanted to do the original ULPC bowtie as well because its slightly larger, and add recolors to that one as well.

So once these things are out of the way, we'll have all of Eliza's assets merged properly and this task can be closed.

Again, I really wanted to focus on migrating Eliza's assets over because she's got the extra animations like run which most other assets don't have. So it greatly cuts back on the time to add something for these additional animations.

**Comment by jrconway3** [comment]
I already mapped the Eyebrows, that was pretty easy to do. I'll work on Facial Hair later, Glasses, and the Eyepatch a bit later. I'll probably open a PR for all these "easy assets" on their own before going in and finishing the clothing.

**Comment by jrconway3** [comment]
Facial Hair is still missing, I'll do that next. I'll map both the LPC Revised facial hair as well as remapping all the existing ULPC facial hair.


Clothing will still take some time unfortunately. I am working on creating mapped "sleeveless" shirts for all original base animations and am creating the sleeves on an overlay, at which point I can then apply the sleeves to _all_ shirts of one type at once, because the sleeves are basically identical between assets. I have most of the t-shirt sleeves done so far except "shoot", but then I need to still do longsleeve as well.

I'll also compare the helms and see where they line up. I think these might be the same helms already here, but they're still missing (obviously) from the newly added animations. And its also possible the helms are the wrong size due to LPC Revised small head sizes.

**Comment by jrconway3** [comment]
So latest PR added facial hair, still need to finish clothes. But I also forgot about helmet assets:
https://github.com/ElizaWy/LPC/tree/main/Characters/Head%20Accessories

And there's a sword and shield, too:
https://github.com/ElizaWy/LPC/tree/main/Characters/Props

That said, I think the helmets might already be here, anyway... Still, this would add the remaining animations. I'll check it later.

**Comment by fnaith** [comment]
At this point, I am finally aware of how much works need to be done.
Are there some simple repetitive works can be done by custom python script? (like replace a clip of image by others )

**Comment by jrconway3** [comment]
Bluecarrot16's lpctools:
https://github.com/bluecarrot16/lpctools

I already use it for handling recolors, however, the problem right now is that I can't get rearrange to work properly. Rearrange is intended to work in such a way that you can position certain assets to all frames and it automatically cuts out the arms for certain assets. However, I tried this with my ear assets and it turned out badly. However, thinking it over, its POSSIBLE that lpctools were configured to map settings for the old v2 body models rather than v3. If that's the case, its not impossible that I simply need to adjust the positioning, but I don't know how to do that yet.

However, if I look into it soon, that might actually help out and make it way easier to map out "easy" assets to all frames. That still won't solve the more complicated assets, though, like clothing, and frankly, there's no real easy way to do clothing. Perhaps I could have it map set clothing assets to all frames and have it cut out the arms, and that might help a lot but it'll still require adjusting clothing assets on a frame-by-frame basis.


That being said, even here it might save a lot of time if I could get rearrange to work. So perhaps I'll look into modifying lpctools soon for rearrange.


Still, other things will always be more difficult. Run and jump animations for assets like hair won't look good simply mapped to all frames and probably need animations. Sleeves on clothing absolutely must be mapped to every frame to fit the arms properly. (Sleeveless clothes aren't NEARLY as hard, as we simply need to map to each body type; at that point mapping to all frames with a few slight body variation positions, then cutting out the arm positions, will probably be fine).

**Comment by fnaith** [comment]
Judge by the [layout](https://github.com/bluecarrot16/lpctools/blob/main/tests/arrange_files/layout/universal.png), I guess the lpctools may not support v3 very well.

For rearranging assets, I think the proper method is assign the positioning manually and let tools to do the work.

But I don't have much knowledge, so it requires some one to describe the detailed mapping method (per asset type).
And if I failed to build such tool, it would be a waste of time.

🤔 🤔 🤔

**Comment by jrconway3** [comment]
I'm going to finally continue working on these clothing assets tonight. Its been a while since I had time to do so. I'll probably spend the next two hours or so working on other asset mapping and then stream doing clothing updates.

**Comment by jrconway3** [comment]
Made a good amount of progress on shirts now:
https://www.youtube.com/watch?v=qdbMXeUJ6OU

I created "sleeveless" versions of Eliza's base female shirt and mapped the alternate base shirt designs to all frames. I also have the short sleeves on a separate layer so I can easily paste the missing short sleeves over all animations to very quickly have 4 different T-Shirt variants that have all animations present. I just need to finish creating the Buttoned and V-Neck shirt positions for the third and fourth "south" shoot animation, and once I have those I can paste them to all frames, then cut out the arms and simply paste the sleeves over them.


Eliza's shirts are similar but slightly different to the base shirts here already, as I mentioned above. There's a basic simple shirt, along with buttoned, v-neck, and scoop, both in T-Shirt and Longsleeve variants. The sleeveless ones I intend on creating a separate asset for as well, but I do need to still clean them up slightly to make them look good enough.

I'm hoping the Teen variant won't be too hard to do, either... The sleeveless cutouts should almost work as is, but I may need to cut out a tad bit more due to the sleeveless female shirts going around the breasts whereas the "teen" / "male" shirts obviously do not in that case. Hopefully the cutouts will do well to cut out the sleeves on the teen variant but I may have to do a bit more tweaking to it.


I'll resume this again tomorrow night most likely.

**Comment by jrconway3** [comment]
Teen/"Thin Male" and Female shirts are now up, both Sleeveless (listed as Sleeveless 2) and T-Shirt. These are in the subdirectory organization I made before where Shortsleeve, Sleeveless, etc, are under separate directories under shirts. Longsleeve coming soon but I need to create sleeves on all frames here first.

I won't do Pregnant/Male just yet but these shouldn't be too bad. For the male assets the shirts just need to be expanded width-wise over many frames but I also need to redraw the sleeves which is probably the biggest issue. However, made "twisted  body" frames are _identical_ which makes this a lot easier.

Pregnant I still need to finish adding the belly to combat poses. Then I need to figure out how existing assets map over to this. This one is probably much harder so I'm more hesitant to do it, but I think once I have a mapping layer over the belly I can then do this for other assets as well such as my kimonos, too. The main difference would be how the shading looks and if I need to add anything extra to make it mesh well with the original asset.


My intent is to clear out Eliza's list of assets first. However because the "easy" assets are relatively easy, I'm also gradually working on head assets which just need to be positioned over the head in various places as well as adding more assets to child frames. The biggest issues are assets which probably should be animated such as longer hairstyles which I'm more hesitant to just jump in and do.

**Comment by jrconway3** [comment]
Started working on helmets today. I basically got all of them, however, the "mail hood" looks way different so I kinda want to map that one as well, but I'd have to redesign it for the larger heads. The basic mail hood also looks off without any animation on the new combat slashes, but I'm not worrying about that right now.

Also for the most part, as I thought, these helmets are smaller than our current heads, but all the helmets already exist on ULPC (except, again, the mail hood which looks a lot different) so I'm just mapping these ones to all frames rather than using the versions from Eliza's repo. I'll probably map all helmets to all frames rather than just the ones Eliza has. Where possible I'll add recolors, but if there's too many shades I won't do recolors (yet). The mail is notable--its got a ton of shades and I have no idea how to properly handle recolors there yet.

The shields look incredibly obnoxious to deal with, there's many varieties of symbol overlays on them so I'm not sure how to deal with them yet. That said, all these sword and shield animations are oversized, so I'll just grab the base oversize animations and simply adjust the slash/backslash to be the same layout as the current one rather than map them to the full layout.

**Comment by jrconway3** [comment]
Ah, apparently the Heater Shield already exists here. So I just need to add the new frames to it. That's cool, that makes things a lot easier.

**Comment by jrconway3** [comment]
Not so cool because once I place this on all frames I need recolors for all variants of patterns. That I feel like I'll wait on... that looks obnoxious lol. I'll just finish the helmets for now and worry about the sword/shield until later.

**Comment by jrconway3** [comment]
Longsleeve shirts are done. Only thing left is the Cardigan now for shirts. I do, however, plan to include some extra assets when I do the Cardigan, notably the Suspenders which are missing a ton of frames.

**Comment by jrconway3** [comment]
Apologies for the delay in continuing this, just been a lot of stuff going on this summer. That said, I've been heavily thinking about trying to go back to doing a couple of things here and there just to slowly make progress. I have also been dabbling in other things lately (started working with a neat romhacking program called FE Builder for a fun project).


Anyway, I was thinking I might finally start working on those damn shields. If I slowly start working on each little piece I can probably get through it. I think the shields should likely be very easy, honestly--Eliza already has the shields mapped to all her assets, and the shields are already mapped to all these assets. So the only thing to do is combine them. The problem isn't how difficult they are, it was intimidating because when bluecarrot ported them over he kept all of Eliza's palettes instead of repaletting. (I have considered maybe porting over Eliza's palettes, too, but that's a lot of extra stuff.)


That means I'll have to create a new palette recolor for all of Eliza's palettes, and furthermore, I'll have to go through each individual shield addon and do it all. That's why I hesitated before. However, shields and the leg assets (along with the Cardigan shirt) are the only things left. I didn't just reuse my old method for Cardigan because the sleeves have a different pattern, which is why it didn't get finished.

**Comment by jrconway3** [comment]
I finished the base Shields, Paint, and Trim last night on stream. I did make some adjustments to bluecarrot's original positioning of the shields, but once I got the trim in place, I ended up restoring some of the old positions anyway.

Right now I just need to implement palette swaps, then I'll have a new PR. My biggest concern is that some of the entries are using Eliza's full list of palette swaps and I want to preserve that, but that also means creating a new palette json file with _all_ of Eliza's palettes, which is super annoying. Honestly it would be nice to just dynamically generate the palettes using JS or something.


That being said, when I do release this I'll probably make it a separate listing for now. That's so the old shield patterns still work. I'll only replace it completely once I finish all shield patterns, and even then, perhaps I'll leave the old one in in case people prefer the old positioning. I haven't decided that yet.

**Comment by jrconway3** [comment]
I think I got the list of LPCR alt palettes about half configured now. The body palettes, metal palettes, and wood palettes are all done, so once I finish the alt palette lists, I can finally generate all the recolors for the paint and push the next PR. The next PR will also include the palette JSON files as well to use in lpctools.

**Comment by jrconway3** [comment]
Okay, PR is now open. This includes only the Heater Shield Base, the Trim/Frames of the shield, and the paint of the shield. I had to add palettes for LPC Revised as well, which were added to `palettes/revised` to seperate them out.

I'll try getting through the patterns relatively quickly, but I'll be mixing up working on this with other things, so it may still take a while to get it all done. I'd like to drop a PR per week with new patterns, but we'll see. I can't guarantee anything.


Once the patterns are done, I have the Arming Sword, Cardigan, leg assets (shorts, pants, etc), and Suspenders to complete--I forgot about suspenders as well. I THINK that's all that's left? But its still a decent chunk left. And even then, this is only assets for Female/Teen, I haven't even made the classic Male versions of these assets or Pregnant, let alone Muscular. These I won't worry about yet before closing this issue, I'll open a new issue later for handling alternate body sizes.

**Comment by jrconway3** [comment]
Looks to me like Bendy, Bendy Sinister, and Lozengy are actually unfinished. The main designs here in both LPC Revised and ULPC are just flat paint. However, some of the frames are finished on LPC Revised. These ones I will definitely hold off on, because its going to take more work to actually finish them, but ideally I'd actually "finish" them and not just use them as is.

Speaking of which, I feel like "Sinister" should be renamed "Inverted". Because all these do is, if the pattern goes in one direction, it flips the direction.

**Comment by jrconway3** [comment]
I think I'll do Chief, Cross, Fess, Pale, Pall, Paly, and Quarterly next. After that I'll do all the "Per" entries.

I'm probably going to have to adjust the shield position on the walk animations, though. It looks awkward (at least in the "north" facing animation). I may just revert this back to the original position. I still prefer putting the shield on the back in some poses, but other changes I made to the positioning doesn't look good, which is why I kept the original positions as a separate asset.

**Comment by jrconway3** [comment]
So in addition to the above patterns, I also configured Lozengy. I did not anticipate doing this yet, but I looked it up and a Lozengy is basically a diamond pattern. So when the shield is tilted I made them squares... trying to recreate the original pattern at different angles is really annoying, though. So if anything looks off here please let me know ASAP.


I also adjusted the north walk position of the shield on my "Revised" animations to make it closer to bluecarrot's original. I didn't think the one I did with the shifting shield positon looked good at all.


I also have been making sure to attach my new frame positions to the original positions as well so that way people can use both if they prefer.

I'll do the "Per" entries next, and maybe at the same time try and finish Bendy and Paly. As it turns out, Paly is ALSO unfinished. I don't think it should be too hard, though. Its basically just erasing pixels in two lines along the middle, so that way there's three columns on this pattern. I'm more worried about Bendy than Paly tbh.

**Comment by jrconway3** [comment]
At this point I have all patterns configured except Bendy and Bendy Sinister, which I'm just finishing up now. To make the ones that are missing full patterns I'm just using the filled out shield paint and erasing it in the pattern, trying to match the original pattern.

Out of all of these, the only one I really had a lot of trouble figuring out is Lozengy. I might redo some frames there as I'm not fully happy with that one.

If I finish these last few assets before the last PR gets merged, I'll just merge these finishing touches into that one.


To finish off this issue, I still have the following:
1. Overalls
2. All Pants/Legs Assets
3. Socks / Shoes / Boots
4. Cardigan (I want to adjust the sleeves to fit the sleeve pattern on the default Cardigan)
5. Wings on Combat Frames (all other wing frames are done)
6. Arming Sword

Still a lot more than I was hoping. Wings is probably the easiest as I just have to figure out how to handle the combat animation shifts on them.

**Comment by jrconway3** [comment]
This PR now finishes off all the remaining Heater Shield patterns:
https://github.com/sanderfrenken/Universal-LPC-Spritesheet-Character-Generator/pull/186

So that's a huge chunk of annoyance out of the way. I'll probably take a break for a short while, but like I said above, I'll probably finish the wings next. I just have to put wings on the combat animations to finish it off. I don't intend on creating new frames, but I'm still not exactly sure which frames to use here.


I really want to finish up this card so it can finally be closed.

**Comment by jrconway3** [comment]
Wings are now mapped to all frames. I'm quite happy with how these turned out. Figuring out the combat animations was quite difficult and I had to add new transition frames for the "Bat" wings.

I've scoped out a bit of a game plan on how to handle the leg assets now. The styling on all the leg assets are loose variations of each other, with different cutoffs. The two styles of pants are identical except one version has a "cuff" on the bottom of the leg, similar to how the Cardigan also has a "cuff" at the end of the sleeves. Those cuffs are the main reason I didn't finish Cardigan yet, but I'll do that alongside the new assets.

The two sets of socks are just cut off versions of the leggings, cut off at different points. The main idea should be, I think, to just port the leggings to all frames, then cut them down to the necessary lengths for each asset. What's also nice about this is I don't have to do it twice, once for teen and once for female, because the legs are the same between both bodies.


Suspenders... I think this shouldn't be too hard either. I have it mapped to most frames now, I just need to cutout the arms. Some frames need the position altered slightly to fit better, but I can do that relatively easily.

And the Overalls have the same base as the regular pants, so I would just need to add in the upper portion of the overalls.

**Comment by jrconway3** [comment]
I put Suspenders under "Jacket" because I felt like that would be the best initial place for it.

I'm going to drop in Cardigan, Suspenders, and Arming Sword into a smaller quick update. I'm mixed on how to handle the sword atm, but I'm thinking it might be best to only have oversize animations for the slash animations, and otherwise only put the sword on the walk animations. The Arming Sword animations are already completed in "oversize", I just have to rename the files and test them. At most I may need to slightly tweak the adjustments

Cardigan is nothing more than putting cuffs on the sleeves, so I'll do that on my own time within the next week.


I plan to stream doing the leg assets, and I think I can get through them relatively quickly. However, I am currently working on Echoes of Wisdom in regards to streaming and I haven't found time between work and other things to work on these assets. Plus, if I did all these updates at once anyway it would be a HUGE update. So I'll have a "misc" update that contains Cardigan, Suspenders, and Arming Sword before the final update with the leg assets.

What's nice about the leg assets is they fit everything except the classic "male" outfits just fine without any changes, so they only have to be done once. I'll do the "thick" male assets at a later date, I think, after I work on a few more personal assets.

**Comment by jrconway3** [comment]
So I fixed up the Arming Sword animations now for the oversized (well, 128) animations, and it looks REALLY good just dropping it right in. Cardigan is just about done as well, so I should be able to get this update out this weekend. I just have to finish getting Arming Sword on "walk" and "down". I don't care about any other frames as I'm basing this on how other swords work, plus, you don't really want to hold a sword while thrusting with two hands, casting a spell, or shooting a bow anyway. Doesn't make much sense. So this should be good.

About the only other thing I can think we might want is perhaps idle animations... I can probably drop those in, too.

Next week(end) I'm planning to stream doing the leg assets.

**Comment by jrconway3** [comment]
It appears the overalls may already be done, so that's intriguing. I didn't notice them in here until now. I'll take a closer look at them once I get around to it. It might make it easier to finish up all the other assets, but we'll see. The Overalls (under "Aprons", which is where I moved Suspenders to) do credit ElizaWy, so perhaps they're the same ones I have yet to finish mapping.

Anyway, I finished Arming Sword, Suspenders, and Cardigan. For the Arming Sword, I just grabbed all the oversized animations from Eliza directly, with minor tweaks. On the walk animation, I based the positioning off of the Rapier, except I made the sword go down on north/south walk instead of go up.

**Comment by jrconway3** [comment]
Overalls aren't quite done, but they're mostly done. I fairly quickly fixed up most of the "Teen" assets. The only thing missing right now are Eliza's combat animations, for whatever reason these don't exist on those assets. The cuffed pants are basically the same lower half, though, so I can grab the pant portion from cuffed pants, and I just need to use a combination of the suspenders and other frames for overalls to finish these up.

**Comment by jrconway3** [comment]
Another update:
The Male assets are mostly done as well. I will need to expand them out or otherwise shift them slightly, but I think I'll include the "Male" in addition to "Female" and "Teen" for Overalls. I can then use the overalls to begin the process of handing the regular pants later, but I won't fully do Male right away there because there's just so much to work on.

I'll probably create a separate issue later to address "Male" and "Pregnant" assets of these clothing pieces from LPC Revised.

**Comment by jrconway3** [comment]
Teen/Female Overalls are done now. I can use these as a starting point for the rest of the pants assets (which is literally just the standard assets before adding run, jump, and eliza's combat animations).

I'll finish Male as well, all I need to do is adjust a few frames here and there for jump, run, climb, etc., and its not even a big change overall either. Only real issue is we lost the lightest color on some of the frames (it got merged into the second lightest color).

Not sure what happened, but basically, the prior set of overalls were missing the darkest non-border color. To keep it consistent, I had to just merge the lightest color from Eliza's into the second-lightest color.

**Comment by jrconway3** [comment]
Okay, just one thing left to go, and that's shorts. I'll finish Shorts by the end of the week, then this issue is officially DONE.

**Comment by jrconway3** [comment]
Shorts are a bit more complicated than I thought. However, at this point Regular Shorts have just the combat animations left, then I'll use the regular shorts to create Short Shorts and we're good.

**Comment by jrconway3** [comment]
Final PR is open now.

I was a bit concerned on a few of the combat animations for short shorts. After putting them into the generator, I didn't like the gap on the other leg, so I opted to put in a small bit there.

The way these shorts work are a bit weird, because they basically don't touch the legs at all. For the running views, when one leg is clearly in front of the other, they're only at the center of the front most leg. I tried that at first, but it didn't look right for a few frames, so I added some tiny touchups to the inner leg.



Also included the smaller heads as a bonus. Many assets will not look good with these, and I'm not about to fix any assets to make them work. I did check, though, and the glasses don't look too bad. So some things _might_ work. Most hairstyles should be fine, but it'll just mean the size of the head is irrelevant in that case, so maybe it wasn't worth including them. But IDK. Maybe some people would prefer trying the smaller heads, and it was an easy extra addition.
